From: Mahmoud Adam <mngyadam@amazon.com>
Date: Fri, 23 Jun 2023 17:08:41 +0000
Subject: crypto: dh - Add SP800-56A rev 3 Pair-wise Consistency check

add pair-wise consistency check in fips mode as per SP800-56 rev 3
section 5.6.2.1.4:

For an FFC key pair (x, y): Use the private key, x,
along with the generator g and prime modulus p included in the domain
parameters associated with the key pair to compute g^x mod p. Compare
the result to the public key, y.

Signed-off-by: Mahmoud Adam <mngyadam@amazon.com>

[The original patch reads 'ret = _compute_val(ctx, ctx->g, y);' in
dh_pairwaise_test - that will overwrite the public key with the value we
intend to check it against. Instead, use the previously allocated val to
hold this value. Also use mpi_cmp which expects two MPIs instead of
mpi_cmp_ui, which expects an MPI and an unsigned long -hailmo ]
---
 crypto/dh.c | 36 ++++++++++++++++++++++++++++++++++++
 1 file changed, 36 insertions(+)

diff --git a/crypto/dh.c b/crypto/dh.c
--- a/crypto/dh.c
+++ b/crypto/dh.c
@@ -163,6 +163,37 @@ static int dh_is_pubkey_valid(struct dh_ctx *ctx, MPI y)
 	return 0;
 }
 
+/*
+ * SP800-56A pair-wise consistency check:
+ * section 5.6.2.1.4:
+ * For an FFC key pair (x, y): Use the private key, x, along with the generator g and
+ * prime modulus p included in the domain parameters associated with the key pair
+ * to compute g^x mod p. Compare the result to the public key, y.
+ */
+static int dh_pairwise_test(struct dh_ctx *ctx, MPI y)
+{
+	int ret;
+	MPI val;
+	val = mpi_alloc(0);
+	if (!val)
+		return -ENOMEM;
+
+	ret = _compute_val(ctx, ctx->g, val); /*g^x mod p*/
+	if (ret) {
+		goto err_free_val;
+	}
+
+	if (mpi_cmp(val, y)) {
+		ret = -EINVAL;
+		goto err_free_val;
+	}
+
+	ret = 0;
+err_free_val:
+	mpi_free(val);
+	return ret;
+}
+
 static int dh_compute_value(struct kpp_request *req)
 {
 	struct crypto_kpp *tfm = crypto_kpp_reqtfm(req);
@@ -226,6 +257,11 @@ static int dh_compute_value(struct kpp_request *req)
 
 		/* SP800-56A rev 3 5.6.2.1.3 key check */
 		} else {
+			if (fips_enabled && dh_pairwise_test(ctx, val)) {
+				ret = -EAGAIN;
+				goto err_free_val;
+			}
+
 			if (dh_is_pubkey_valid(ctx, val)) {
 				ret = -EAGAIN;
 				goto err_free_val;

